modules/python3/bindings/src/pydatareaders.cpp

```diff
@@ -0,0 +1,94 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2017-2021 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwopy/pydatareaders.h>
+#include <inviwopy/vectoridentifierwrapper.h>
+#include <inviwopy/pypropertytypehook.h>
+
+#include <inviwo/core/io/datareader.h>
+#include <inviwo/core/io/datareaderfactory.h>
+#include <inviwo/core/common/inviwoapplication.h>
+#include <inviwo/core/datastructures/geometry/mesh.h>
+#include <inviwo/core/datastructures/image/image.h>
+#include <inviwo/core/datastructures/volume/volume.h>
+
+#include <warn/push>
+#include <warn/ignore/shadow>
+#include <pybind11/operators.h>
+#include <pybind11/stl.h>
+#include <warn/pop>
+
+namespace inviwo {
+
+
+void exposeDataReaders(pybind11::module& m) {
+    namespace py = pybind11;
+
+    py::class_<DataReader>(m, "DataReader")
+        .def("clone", &DataReader::clone)
+        .def("getExtensions", &DataReader::getExtensions)
+        .def("addExtension", &DataReader::addExtension)
+        .def("setOption", &DataReader::setOption)
+        .def("getOption", &DataReader::getOption);
+
+    py::class_<DataReaderType<Image>, DataReader>(m, "ImageDataReader")
+        .def("readData", [](DataReaderType<Image>* reader, std::string_view filePath) {
+                return reader->readData(filePath);
+            },
+            py::return_value_policy::reference);
```

Comment:
explain why the return value should be reference? seems very strange. 
Why no init function? 
We should have a trampoline class here to

Reply:
> explain why the return value should be reference? seems very strange.

Fixed. 

> Why no init function?

They are abstract classes.

> We should have a trampoline class here to

Yes, it seems like many of the exposed classes are missing trampolines so it won't be possible to extend those classes in Python. The Port is one such example.
I have added trampolines for the DataReader and DataReaderType.